src/main/java/org/springframework/integration/aws/inbound/kinesis/KinesisMessageDrivenChannelAdapter.java

```diff
@@ -1006,7 +1018,23 @@ private Runnable processTask() {
 								KinesisMessageDrivenChannelAdapter.this.shardConsumerManager.shardOffsetsToConsumer
 										.remove(this.key);
 							}
+
 							// Shard is closed: nothing to consume any more.
+							// Checkpoint endingSequenceNumber.
+							if (KinesisMessageDrivenChannelAdapter.this.checkpointClosedShards) {
+								for (Shard shard : readShardList(this.shardOffset.getStream())) {
+									if (shard.getShardId().equals(shardOffset.getShard())) {
+										String endingSequenceNumber = shard.getSequenceNumberRange().getEndingSequenceNumber();
+										if (endingSequenceNumber != null) {
+											checkpointer.setHighestSequence(endingSequenceNumber);
+											checkpointIfBatchMode();
+											checkpointIfPeriodicMode(null);
+											checkpointIfRecordMode(null);
+										}
```

Comment:
I think this (closed shard) is a special use-case and we should call `checkpointer.checkpoint()` independently of the mode, since this (closed shard) is not going to effect end-user at all.

Not sure so that we have some `endingSequenceNumber`, but I guess just plain `checkpointer.checkpoint(endingSequenceNumber)` is fully enough to satisfy what you would expect from the closed shard handling.

Reply:
My concern is when in `manual` checkpoint mode.  Any other mode then it's fine to do as you say.

But `manual` is a bit of an exception as the user has decided to take complete responsibility for marking a record as "done".

If we automatically checkpoint on stream-end here, we are effectively marking all records since the last checkpoint as done. 

Imagine the user has chosen `manual` checkpointing because they want guaranteed at-least-once semantics, and will only checkpoint once they have positive confirmation that a record is read _and processed_ - if an exception or system termination occurs since the last (manual) checkpoint, they would be happy to retry and reprocess the record.  But it is not acceptable to skip it.

However, if we checkpoint at stream-end here, there could be a whole queue of records waiting to be processed, and if something goes wrong and the system restarts, those queued items will not be retired.

This is especially the case if there is an actual queue implemented with another thread draining it (for some performance reason, so downstream record processing doesn't block the kinesis reader thread).  

Does that make sense?

Reply:
I have pushed some changes, but I am now checking if we are in `manual` mode and *not* checkpointing if we are.

To be clear, I am not actually using `manual` checkpoint mode in my projects, so I don't feel strongly about this.  If you want I can change my PR to always checkpoint, regardless of mode 😄 

However, I do think to do _any_ automatic checkpointing at all when in `manual` mode would be a bug... it's your call! 😅

Reply:
Yes, I think you have a point when we really have some records in the `GetRecordsResult` to process, when data really reaches end-user code for that possible manual checkpointing. But when it is empty like in your case, we have to checkpoint it automatically on the framework level.

Makes sense?

Reply:
Hi, sorry, I think I have confused myself slightly here 🤦 

Let me step back. Please bear with me...

I actually think there is a slightly different, bigger bug here, I'm sorry I've made a bit of a mess of this 😆 

So, I have just empirically confirmed the following: the `endingSeqenceNumber` of a closed shard *is not* the same as the sequence number of the last record written to that shard.  In fact  the `endingSeqenceNumber` is larger (as you probably guessed).

It looks like the code in master at the moment is not actually ever checkpointing the `endingSeqenceNumber`, at least I don't see any calls to `getEndingSequenceNumber()` that would do that 🤔 .  At most, it will checkpoint the sequence number of the _latest record_.

So, as far as I can see, currently the code has a bug where closed shards _will never be "skipped"_.  This includes non-empty shards.

I am testing this in `us-east-1` and I seem to have positive confirmation of this in a development instance I am working with where I have just resharded.

In both cases, we will see repeated log INFO messages like
`Stopping the [ShardConsumer{shardOffset=KinesisShardOffset{iteratorType=AFTER_SEQUENCE_NUMBER, sequenceNumber='49610015252819166836042076049393253537413313796583195858', timestamp=null, stream='test-stream', shard='shardId-000000000077', reset=false}, state=STOP}] on the checkpoint [49610015252819166836042076049393253537413313796583195858] because the shard has been CLOSED and exhausted.`

The difference is that empty shards will have `iteratorType=LATEST`

Anyway, the conclusion is that we should always checkpoint the `endingSequenceNumber` on stream-end, unless we are in `manual` mode, and the tests should be updated to have a different `endingSequenceNumber` to the last record's sequence number 😄

Reply:
Updated tests pushed

Reply:
So, sounds like we are on the same page. It is indeed a missed functionality on the master. That's why we have a fix like this from you.
Only the point to checkpoint properly and probably you are right about this `endingSequenceNumber`. Only need to agree about auto-checkpinting when we are in the manual mode and result is empty.

Reply:
Okay, I now do the checkpoint if 
* We are not in manual mode
or
* `this.checkpointer.getLastCheckpointValue() == null` - this should be true only for empty shards